File: c++ version/500+/739-daily-temperatures.cpp
vector<int> dailyTemperatures(vector<int>& temperatures) {
    int n = temperatures.size();
    vector<int> ans(n, 0);
    stack<int> s;
    for (int i = 0; i < n; ++i) {
        while (!s.empty() && temperatures[s.top()] < temperatures[i]) {
            int pre = s.top();
            s.pop();
            ans[pre] = i - pre;
        }
        s.push(i);
    }
    return ans;
}

File: python version/739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        ans = [0] * n
        stack = []
        for i in range(n):
            while stack and temperatures[stack[-1]] < temperatures[i]:
                prev = stack.pop(-1)
                duration = i - prev
                ans[prev] = duration
            stack.append(i)

        return ans